Update kafka-clients to 2.6

pskianis-sky opened the pull request:
* Had to increase the Tolerance for some tests

paulaylingdev commented on build.sbt:
```diff
@@ -43,7 +43,7 @@ val dependencies = Seq(
   "org.mockito"                % "mockito-all"                  % "1.10.19" % Test,
   "org.zalando"                %% "grafter"                     % "2.6.1" % Test,
-  "io.github.embeddedkafka"    %% "embedded-kafka"              % "2.2.0" % Test
+  "io.github.embeddedkafka"    %% "embedded-kafka"              % "2.6.0" % Test
```
```suggestion
  "io.github.embeddedkafka"    %% "embedded-kafka"              % kafkaVersion % Test
```

This library usually follows the kafka version closely so I think it is fine to share the same variable

pskianis-sky replied:
It does not always do that, especially with patches. so I will leave as is unless you feel strongly about it.

paulaylingdev commented on scheduler/src/test/scala/com/sky/kms/base/KafkaIntSpecBase.scala:
```diff
@@ -4,26 +4,29 @@ import com.sky.kms.kafka.Topic
-    withConsumer[Array[Byte], T, ConsumerRecord[Array[Byte], T]](_.consumeLazily(topic)(identity, retryConfig).head)
+    withConsumer[Array[Byte], T, ConsumerRecord[Array[Byte], T]] { cr =>
+      cr.subscribe(List(topic).asJavaCollection)
+      cr.poll(JDuration.ofSeconds(60)).iterator().next()
```
Just to give the 60 a name how about something like:

```scala
  def consumeFirstFrom[T : Deserializer](topic: String, timeout: Duration = 60.seconds): ConsumerRecord[Array[Byte], T] =
    withConsumer[Array[Byte], T, ConsumerRecord[Array[Byte], T]] { cr =>
      cr.subscribe(List(topic).asJavaCollection)
      cr.poll(timeout).iterator().next()
```

Same for the `consumeSomeFrom` function

pskianis-sky replied:
Done

paulaylingdev reviewed (commented):
Minor comments but looks good thanks for the contribution!

paulaylingdev commented on scheduler/src/test/scala/com/sky/kms/base/KafkaIntSpecBase.scala:
```diff
@@ -4,26 +4,35 @@ import com.sky.kms.kafka.Topic
     withConsumer { cr: KafkaConsumer[String, T] =>
-      cr.consumeLazily(topic)(identity, retryConfig).take(numMsgs).toList
+      cr.subscribe(List(topic).asJavaCollection)
+      cr.poll(JDuration.ofSeconds(timeout.toSeconds)).iterator().asScala.toList.take(numMsgs)
```
```suggestion
      cr.poll(timeout.toJava).iterator().asScala.toList.take(numMsgs)
```

Then the java import can be removed

pskianis-sky replied:
oops, missed that one, one before is doing exactly that. thx

peter-hazell commented on scheduler/src/test/scala/com/sky/kms/base/KafkaIntSpecBase.scala:
```diff
@@ -4,26 +4,34 @@ import com.sky.kms.kafka.Topic
+  def consumeFirstFrom[T : Deserializer](topic: String,
+                                         timeout: FiniteDuration = 60.seconds): ConsumerRecord[Array[Byte], T] =
+    withConsumer[Array[Byte], T, ConsumerRecord[Array[Byte], T]] { cr =>
+      cr.subscribe(List(topic).asJavaCollection)
```
Maybe we could generalise subscribe and poll so that we don't duplicate it in here:

E.g:

```
def subscribeAndPoll[K, V](
      topic: String,
      timeout: FiniteDuration = 60.second): KafkaConsumer[K, V] => Iterator[ConsumerRecord[K, V]] = { cr =>
    cr.subscribe(List(topic).asJavaCollection)
    cr.poll(timeout.toJava).iterator().asScala
  }
```

and then we can do:

`subscribeAndPoll(topic, timeout)(cr).toList.take(numMsgs)`

and

`subscribeAndPoll(topic, timeout)(cr).next()`

paoloambrosio-skyuk replied:
It is a shame that Embedded Kafka doesn't come with methods that return a `ConsumerRecord`. We should really try and add that do the library so that we can remove this altogether.

pskianis-sky replied:
Applied the subscribeAndPoll recommendation.

lacarvalho91 replied:
im sure Embedded Kafka had methods that return `ConsumerRecord`...wonder if they were removed? :S i'll have a look

lacarvalho91 replied:
bah...yeah deprecated

paoloambrosio-skyuk replied:
`ConsumerRecord ` is removed in the lower layer, so there isn't much we can do there unless we want to send a PR.

paoloambrosio-skyuk commented on scheduler/src/test/scala/com/sky/kms/base/KafkaIntSpecBase.scala:
```diff
@@ -4,26 +4,39 @@ import com.sky.kms.kafka.Topic
+import scala.concurrent.duration._
+import scala.compat.java8.DurationConverters._
+
+trait KafkaIntSpecBase extends EmbeddedKafka with WordSpecLike {
```
```suggestion
trait KafkaIntSpecBase extends EmbeddedKafka { this: PatienceConfiguration =>
```

`WordSpecLike` is not needed and I'd rather introduce `PatienceConfiguration` than having a copy/paste of `timeout: FiniteDuration = 60.seconds` all over.

paoloambrosio-skyuk commented on scheduler/src/test/scala/com/sky/kms/e2e/package.scala:
```diff
@@ -8,7 +8,7 @@ import scala.concurrent.duration._
 package object e2e {
 
-  val Tolerance = 400 millis
+  val Tolerance = 1300 millis
```
It fails on my laptop as well... perhaps we could redefine this in terms of PatienceConfig in `SchedulerIntSpecBase`.

paoloambrosio-skyuk reviewed (commented):
Had a couple of comments that touched too many places to be a suggestion. I pushed two commits (not signed 🤦🏻 ) on the `pr99-suggestions` branch. Feel free to cherry pick them or reimplement the changes if people agree with my two points.

pskianis-sky commented:
Partially applied the changes discussed above, added a pollTimeout in kafkaConfig and removed it from the function signatures, and moved the tolerance definition from the e2e package to SchedulerIntSpecBase.

lacarvalho91 commented on scheduler/src/test/scala/com/sky/kms/base/KafkaIntSpecBase.scala:
```diff
@@ -4,26 +4,35 @@ import com.sky.kms.kafka.Topic
 
   val scheduleTopic: Topic      = "scheduleTopic"
   val extraScheduleTopic: Topic = "extraScheduleTopic"
+  val pollTimeout               = 60 seconds
```
```suggestion
  private val pollTimeout               = 60 seconds
```

pskianis-sky replied:
wouldn't you want that to be overridable (in general)

paoloambrosio-skyuk replied:
I'd be happy with either `private val` or `def` (that can be overridden with a val or a def). If public, I'd like to see a name that expresses that it's a timeout for Kafka (e.g. `kafkaTimeout`, `kafkaOperationsTimeout`, ... [1]) otherwise in the overriding class you wouldn't be able to tell immediately what timeout you are configuring without finding the declaring class.

[1]: I find that using the word "poll" breaks encapsulation because externally clients of this interface shouldn't care about polling or how we consume those messages, but this is a detail. Even `subscribeAndPoll` should probably be made private if it's used only internally.

pskianis-sky replied:
* polITimeout => kafkaConsumerTimeout
   Originally called pollTimeout because the function was subscribeAndPoll, the kafka client's function is poll and Kafka is in the name of the base class. However, I agree it is too generic so I decided to use kafkaConsumerTimeout as there are other (consumer and/or producer) related timeouts as well that could be set up as well as the class just being called KafkaIntSpecBase. Let me know if that's ok.
* prefer the def over the private val or val.
* made subscribeAndPoll private

lacarvalho91 commented on scheduler/src/test/scala/com/sky/kms/base/SchedulerIntSpecBase.scala:
```diff
@@ -2,7 +2,9 @@ package com.sky.kms.base
 
 abstract class SchedulerIntSpecBase extends AkkaKafkaSpecBase {
   implicit val conf = TestConfig(NonEmptyList.of(scheduleTopic, extraScheduleTopic))
+  val Tolerance     = 1300 milliseconds
```
```suggestion
  val tolerance     = 1300 milliseconds
```

Scala [naming convention](https://docs.scala-lang.org/style/naming-conventions.html#constants-values-variable-and-methods) is that only constants defined in objects should be in upper camel case

pskianis-sky replied:
will update

lacarvalho91 reviewed (commented):
almost there, minor changes
